gh-118469: Document sqlite3.Binary (#122343)
python/cpython: Doc/library/sqlite3.rst

      and *uri* is deprecated.
      They will become keyword-only parameters in Python 3.15.

.. function:: Binary(data)

   Create a Binary object to handle binary data in SQLite.

>    :param data:
>        The binary data to be encapsulated. This can be any object supporting the
>        Python buffer protocol, such as bytes, bytearray, or memoryview.
>    :type data: buffer-compatible object

Reviewer: ```suggestion
   :param data:
       The binary data to be encapsulated. This should be a :term:`bytes-like object`.
```

   - Ensures data is bound as a BLOB in SQL queries.

   **Use Cases:**

   - **Storing Raw Binary Data**: Ideal for files, images, or any raw binary data.
   - **Preserving Exact Byte Sequences**: Prevents encoding/decoding errors.
   - **Working with Non-Text Data**: Ensures data isn't subjected to text-based operations.

   **Behavior in Queries:**

   - **INSERT or UPDATE**: Stores Binary objects as BLOB data.
   - **SELECT**: Retrieves BLOB data as Binary objects, convertible to bytes.

   **Performance Considerations:**

   - Using Binary() for large data can be memory-intensive.
   - For very large objects, consider using SQLite's incremental or BLOB I/O interfaces.

   **Compatibility and Portability:**

   - Ensures consistent behavior across Python implementations and SQLite versions.
   - Helps in writing portable code for binary data handling.

   **Error Handling:**

   - Raises :exc:`TypeError` if input doesn't support the buffer protocol.
   - Handle potential :exc:`MemoryError` for very large binary objects.

   **Best Practices:**

   - Use Binary() to explicitly mark data as binary.
   - Be mindful of memory usage for large binary objects.

>    - Consider using with prepared statements for efficiency.

Reviewer: I don't think the following sections should be included. The only one that could be relevant is about "Using Binary() for large data can be memory-intensive", and error handling though I would reformulate the section by saying that it raises a `TypeError` if *data* is of incorrect type. Users should expect `MemoryError` in general (also you say "handle" but how is it handled? it's just re-raising it or?)

   By using :func:`Binary`, developers can ensure robust handling of binary data in SQLite,
   preventing data corruption and enhancing application reliability.

   **Examples:**

   Basic example of storing and retrieving binary data:

   .. code-block:: python

>        import sqlite3

Reviewer: Indent the code according to `code-block`:

```rst
.. code-block:: python
   
$$$import sqlite3
```

or use `::`:

```rst
   Basic example of storing and retrieving binary data::

   $$$$import sqlite3
```

(I used '$' for whitespaces).

       # Create a connection and cursor
       conn = sqlite3.connect(':memory:')
       cursor = conn.cursor()

       # Create a table
       cursor.execute('''
       CREATE TABLE files (
           id INTEGER PRIMARY KEY,
           name TEXT,
           data BLOB

>        )

Reviewer: For a simple example, no need for an ID IMO. Just (name, data)  is sufficient I think.

       ''')

       # Insert binary data

>        with open('example.png', 'rb') as file:

Reviewer: In examples, I'd prefer *not* to create myself an additional file just to see how it could work. Just create some binary data yourself, like `bytes("abc", "utf-8")`.

           binary_data = file.read()

>            cursor.execute('INSERT INTO files (name, data) VALUES (?, ?)', ('example.png', sqlite3.Binary(binary_data)))

Reviewer: Split the call.

       conn.commit()

       # Retrieve binary data
       cursor.execute('SELECT data FROM files WHERE name=?', ('example.png',))
       blob_data = cursor.fetchone()[0]

       # Save retrieved data to a file
       with open('retrieved_example.png', 'wb') as file:

>            file.write(blob_data)

Reviewer: Instead of doing this, you can simply use an `assert` checking that the data being retrieved is the same as the content that was written.

       # Clean up

>        conn.close()

Reviewer: use with statements instead of using `conn.close()`

   Example with buffer-compatible objects:

   .. code-block:: python

       import sqlite3

>        import numpy as np

Reviewer: Can you avoid an example with numpy?

Author: should i keep it or remove it, as u say

Reviewer: If you can have an example without numpy, I think you should use an example based on `io.BytesIO` for instance.

Author: lets remove it :)

       # Create a connection and cursor
       conn = sqlite3.connect(':memory:')
       cursor = conn.cursor()

       # Create a table
       cursor.execute('''
       CREATE TABLE binary_data (
           id INTEGER PRIMARY KEY,
           data BLOB
       )
       ''')

       # Use a memoryview object
       data_array = np.array([1, 2, 3, 4, 5], dtype=np.uint8)
       memoryview_data = memoryview(data_array)

       # Insert memoryview data

>        cursor.execute('INSERT INTO binary_data (data) VALUES (?)', (sqlite3.Binary(memoryview_data),))

Reviewer: Ditto

       conn.commit()

       # Retrieve memoryview data
       cursor.execute('SELECT data FROM binary_data WHERE id=1')
       blob_data = cursor.fetchone()[0]

       # Convert retrieved data back to a numpy array
       retrieved_array = np.frombuffer(blob_data, dtype=np.uint8)

       # Clean up

>        conn.close()

Reviewer: Ditto

      and *uri* is deprecated.
      They will become keyword-only parameters in Python 3.15.

.. function:: Binary(data)

   Create a Binary object to handle binary data in SQLite.

   :param data:
       The binary data to be encapsulated. This should be a 
   :term:`bytes-like object`.

   The :func:`Binary` function encapsulates binary data to ensure proper handling
   by SQLite. It is used to signal to the sqlite3 module that the data should be
   treated as a BLOB (Binary Large Object) rather than text.

   By using :func:`Binary`, developers can ensure robust handling of binary data in SQLite,
   preventing data corruption and enhancing application reliability.

>    **Examples:**

Reviewer: Now it's a single example, so you just write something like:

```rst
Basic example of storing and retrieving binary data:
```

Also please dedent by 1 character the entire code-block (so that the `import` is aligned with `code-block`.

   Basic example of storing and retrieving binary data:

   .. code-block:: python

       import sqlite3

       # Create a connection and cursor

>        conn = sqlite3.connect(':memory:')

Reviewer: Use a [context manager](https://docs.python.org/3/library/sqlite3.html#how-to-use-the-connection-context-manager), you won't need to have `conn.commit` (but you'll need the `.close()` call).

Author: if i use context manager, i dont think i need conn.close() right?

Reviewer: According to what I linked:

```py
# Connection object used as context manager only commits or rollbacks transactions,
# so the connection object should be closed manually
con.close()
```

       cursor = conn.cursor()

       # Create a table
       cursor.execute('''
       CREATE TABLE files (
           name TEXT,
           data BLOB
       )
       ''')

       # Insert binary data

>        binary_data = bytes("abc", "utf-8")

Reviewer: Ok, by example might be a bit too trivial and the way to write it is a bit too verbose, so let's use `binary_data = pickle.dumps({'foo': 42, 'bar': 1337})`. In general, we are interested in doing those kind of things (namely, serializing large objects without bothering about columns).

       query = 'INSERT INTO files (name, data) VALUES (?, ?)'
       values = ('example.txt', sqlite3.Binary(binary_data))
       cursor.execute(query, values)

       conn.commit()

       # Retrieve binary data
       cursor.execute('SELECT data FROM files WHERE name=?', ('example.txt',))
       blob_data = cursor.fetchone()[0]

       with open('retrieved_example.txt', 'wb') as file:

>             file.write(blob_data)

Reviewer: Don't write. People don't like having other files being written. You might also overwrite a local file of the user. Just do something like:

```py
deserialized = pickle.loads(blob_data)
print(deserialized)  # {'foo': 42, 'bar': 1337}
```

(I think this should do something like that)

> 

Reviewer: I'm sorry but the `close()` was actually needed. I thought that the connection would automatically be closed but it's only the commits that are done automatically.

See: https://docs.python.org/3/library/sqlite3.html#how-to-use-the-connection-context-manager

Or you can just put your `.close()` again and ignore my comment on context managers.

   Create a Binary object to handle binary data in SQLite.

   :param data:
      The binary data to be encapsulated. This should be a 

>    :term:`bytes-like object`.

Reviewer: ```suggestion
       The binary data to be encapsulated. This should be a
       :term:`bytes-like object`.
```

Reviewer: The lint error is here by the way. Be careful of spaces.

Reviewer: (And indents, it's really hard for me to do it with GH in comments sometimes because I don't see my own whitespaces).

To fix the lint:

- C/C previous usage (above in the file)
- Change the content.

   By using :func:`Binary`, developers can ensure robust handling of binary data in SQLite,
   preventing data corruption and enhancing application reliability.

>    **Basic example of storing and retrieving binary data:**

Reviewer: If you want to use a title-like header, use `rubric`:

```suggestion
   .. rubric:: Basic example of storing and retrieving binary data
```

or simply write "The following code snippet is a basic example of storing and retrieving binary data:"

   .. code-block:: python

>       import sqlite3
>       import pickle
> 
>       # Create a connection and cursor
>       with sqlite3.connect(':memory:') as conn:
>          cursor = conn.cursor()
> 
>       # Create a table
>       cursor.execute('''
>       CREATE TABLE files (
>          id INTEGER PRIMARY KEY,
>          name TEXT,
>          data BLOB)
>       ''')
> 
>       # Insert binary data
>       binary_data = pickle.dumps({'foo': 42, 'bar': 1337})
>       query = 'INSERT INTO files (name, data) VALUES (?, ?)'
>       values = ('example.pkl', sqlite3.Binary(binary_data))
>       cursor.execute(query, values)
> 
>       # Retrieve binary data
>       cursor.execute('SELECT data FROM files WHERE name=?', ('example.pkl',))
>       blob_data = cursor.fetchone()[0]
> 
>       # Deserialize the binary data using pickle
>       retrieved_data = pickle.loads(blob_data)
>       print(retrieved_data)  # This should print: {'foo': 42, 'bar': 1337}
> 
>       # Save retrieved data to a file
>       with open('retrieved_example.txt', 'wb') as file:
>          file.write(blob_data)
> 
>       conn.close()

Reviewer: ```suggestion
      import pickle
      import sqlite3

      # create a connection and cursor
      conn = sqlite3.connect(':memory:')
      cursor = conn.cursor()

      # create a table
      cursor.execute('''
      CREATE TABLE files (
          id INTEGER PRIMARY KEY,
          name TEXT,
          data BLOB)
      ''')

      # insert some binary data
      binary_data = pickle.dumps({'foo': 42, 'bar': 1337})
      values = ('example.pkl', sqlite3.Binary(binary_data))
      cursor.execute('INSERT INTO files (name, data) VALUES (?, ?)', values)
      conn.commit()

      # retrieve the binary data that was just inserted
      cursor.execute('SELECT data FROM files WHERE name=?', ('example.pkl',))
      blob_data = cursor.fetchone()[0]
      conn.close()

      # deserialize the binary data
      retrieved_data = pickle.loads(blob_data)
      print(retrieved_data)  # This should print: {'foo': 42, 'bar': 1337}
```

Something like that is probably better for the user.
